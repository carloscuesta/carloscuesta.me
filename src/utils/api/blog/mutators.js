// @flow
import { format, formatDistanceToNow } from 'date-fns'
import readingTime from 'reading-time'

export type Post = {
  dateModified: string,
  datePublished: { formatInWords: string, formatDate: string, value: string },
  disqusIdentifier: string,
  excerpt: string,
  html: string,
  images: { featured: { src: string }, preview: { lqpi: string, src: string } },
  readingTime: string,
  slug: string,
  title: string
}

export type PostPreview = $Diff<
  Post,
  {
    dateModified: string,
    disqusIdentifier: string,
    html: string,
    readingTime: string
  }
>

type Payload = { data: Object, html: Object, slug: string }

export const transformPost = (payload: Payload): Post => ({
  dateModified: payload.data.dateModified,
  datePublished: {
    formatInWords: formatDistanceToNow(
      new Date(payload.data.datePublished + ' GMT+2'),
      { addSuffix: true }
    ),
    formatDate: format(
      new Date(payload.data.datePublished),
      'dd MMMM y'
    ),
    value: payload.data.datePublished
  },
  disqusIdentifier: payload.data.disqusIdentifier,
  excerpt: payload.data.excerpt,
  html: payload.html.toString(),
  images: {
    featured: {
      src: payload.data.image
    },
    preview: {
      src: payload.data.image.replace('/upload/', '/upload/w_500/'),
      lqpi: payload.data.image.replace('/upload/', '/upload/t_post-preview-lqpi/')
    }
  },
  readingTime: readingTime(payload.html.toString()).text,
  slug: payload.slug,
  title: payload.data.title
})
