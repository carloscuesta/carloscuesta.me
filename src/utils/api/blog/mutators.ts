import { format, formatDistanceToNow } from 'date-fns'
import { getPlaiceholder } from 'plaiceholder'
import readingTime from 'reading-time'
import type { VFile } from 'vfile'

export type Post = {
  dateModified: string
  datePublished: {
    formatInWords: string
    formatDate: string
    formatMonthDay: string
    value: string
  }
  disqusIdentifier: string
  excerpt: string
  html: string
  images: { featured: { src: string }; preview: { lqpi: string; src: string } }
  readingTime: string
  slug: string
  title: string
}

export type PostPreview = Omit<
  Post,
  'dateModified' | 'disqusIdentifier' | 'html' | 'readingTime'
>

type Payload = {
  data: { [key: string]: string }
  html: VFile
  slug: string
}

export const transformPost = async (payload: Payload): Promise<Post> => {
  const { base64 } = await getPlaiceholder(
    payload.data.image.replace(
      '/upload/',
      '/upload/t_post-preview-lqpi-background-color/',
    ),
  )

  return {
    dateModified: payload.data.dateModified,
    datePublished: {
      formatInWords: formatDistanceToNow(
        new Date(payload.data.datePublished + ' GMT+2'),
        { addSuffix: true },
      ),
      formatDate: format(new Date(payload.data.datePublished), 'dd MMMM y'),
      formatMonthDay: format(new Date(payload.data.datePublished), 'MMM dd'),
      value: payload.data.datePublished,
    },
    disqusIdentifier: payload.data.disqusIdentifier,
    excerpt: payload.data.excerpt,
    html: payload.html.toString(),
    images: {
      featured: {
        src: payload.data.image,
      },
      preview: {
        src: payload.data.image.replace('/upload/', '/upload/w_500/'),
        lqpi: base64,
      },
    },
    readingTime: readingTime(payload.html.toString()).text,
    slug: payload.slug,
    title: payload.data.title,
  }
}
