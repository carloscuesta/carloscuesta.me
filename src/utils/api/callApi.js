// @flow
export type callApiOptions = {
  mutator: Function,
  requestOptions?: Object,
  url: string
}

const callApi = async (options: callApiOptions): Promise<Object> => {
  try {
    const response = await fetch(options.url, options.requestOptions)
    const data = await response.json()
    const dataTransformed = options.mutator(data)

    return dataTransformed
  } catch (exception) {
    console.error('ERROR @ callApi:', exception)
  }
}

export default callApi
