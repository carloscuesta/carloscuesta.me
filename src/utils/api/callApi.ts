export type callApiOptions = {
  mutator: (data: any) => any,
  requestOptions?: RequestInit,
  url: string
}

const callApi = async (options: callApiOptions): Promise<any> => {
  try {
    const response = await fetch(options.url, options.requestOptions)
    const data = await response.json()
    const dataTransformed = options.mutator(data)

    return dataTransformed
  } catch (exception) {
    console.error('ERROR @ callApi:', exception)
  }
}

export default callApi
