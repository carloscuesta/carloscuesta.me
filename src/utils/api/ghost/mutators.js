// @flow
import stripTags from 'striptags'
import moment from 'moment'

type Post = {
  excerpt: string,
  image: string,
  publishedAt: { format: string, value: string },
  slug: string,
  tags: Array<string>,
  title: string
}

export const transformPosts = (payload: Object): Array<Post> => payload.posts.map((post) => ({
  excerpt: `${stripTags(post.html).substring(0, 120)}...`,
  image: post.feature_image.replace('/upload/', '/upload/w_500/'),
  publishedAt: {
    format: moment(post.published_at).startOf('hour').fromNow(),
    value: post.published_at
  },
  slug: `/blog/${post.slug}`,
  tags: post.tags.map((tag) => tag.name),
  title: post.title
}))
