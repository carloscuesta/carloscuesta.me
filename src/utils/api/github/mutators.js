// @flow
export type Repository = {
  description: string,
  language?: string,
  name: string,
  stars: number,
  url: string
}

export const transformRepositories = (payload: Array<Object>): Array<Repository> => {
  return payload.map((repo) => ({
    description: repo.description,
    language: repo.language,
    name: repo.name,
    stars: repo.stargazers_count,
    url: repo.html_url
  }))
}

export type UserInformation = {
  followers: number,
  repositories: number
}

export const transformUserInformation = (payload: Object): UserInformation => ({
  followers: payload.followers,
  repositories: payload.public_repos
})
