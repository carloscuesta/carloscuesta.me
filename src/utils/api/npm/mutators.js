// @flow
export const transformPublishedPackages = (payload: { objects: Array<Object> }): Array<string> => {
  return payload.objects.map((pkg) => pkg.package.name)
}

export const transformDownloadsCount = (payload: Object): number => {
  const packages: Array<Object> = Object.values(payload)

  return packages
    .map((pkg) => pkg.downloads)
    .reduce((memo, value) => memo + value, 0)
}
