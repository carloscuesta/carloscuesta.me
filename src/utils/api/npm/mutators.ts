export const transformPublishedPackages = (
  payload: { objects: Array<{ package: { name: string } }>
}) => {
  return payload.objects.map((pkg) => pkg.package.name)
}

export const transformDownloadsCount = (
  payload: { [key: string]: { downloads: number } }
) => {
  const packages = Object.values(payload)

  return packages
    .map((pkg) => pkg.downloads)
    .reduce((memo, value) => memo + value, 0)
}
