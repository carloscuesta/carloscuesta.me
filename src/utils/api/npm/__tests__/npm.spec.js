import callApi from 'src/utils/api/callApi'
import { fetchPublishedPackages, fetchDownloadsCount } from '../index'
import { transformPublishedPackages, transformDownloadsCount } from '../mutators'

import publishedPackagesFixture from './fixtures/publishedPackages'
import packageDownloadsFixture from './fixtures/downloads'

jest.mock('src/utils/api/callApi')

describe('NPM API Client', () => {
  describe('apiCalls', () => {
    beforeEach(() => {
      callApi.mockReset()
    })

    it('should match fetchPublishedPackages call', () => {
      fetchPublishedPackages()

      expect(callApi.mock.calls[0][0]).toMatchSnapshot()
    })

    it('should match fetchDownloadsCount call', () => {
      fetchDownloadsCount(publishedPackagesFixture.objects.map((pkg) => pkg.name))

      expect(callApi.mock.calls[0][0]).toMatchSnapshot()
    })
  })

  describe('mutators', () => {
    describe('transformPublishedPackages', () => {
      it('should match the packages names', () => {
        expect(transformPublishedPackages(publishedPackagesFixture)).toMatchSnapshot()
      })
    })

    describe('transformDownloadsCount', () => {
      it('should aggregate all the downloads from packages', () => {
        expect(transformDownloadsCount(packageDownloadsFixture)).toMatchSnapshot()
      })
    })
  })
})
