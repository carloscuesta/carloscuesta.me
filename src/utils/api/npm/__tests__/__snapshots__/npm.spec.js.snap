// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NPM API Client apiCalls should match fetchDownloadsCount call 1`] = `
{
  "mutator": [Function],
  "url": "https://api.npmjs.org/downloads/point/last-year/,,,,,",
}
`;

exports[`NPM API Client apiCalls should match fetchPublishedPackages call 1`] = `
{
  "mutator": [Function],
  "url": "https://registry.npmjs.org/-/v1/search?text=author:carloscuesta",
}
`;

exports[`NPM API Client mutators transformDownloadsCount should aggregate all the downloads from packages 1`] = `151010`;

exports[`NPM API Client mutators transformPublishedPackages should match the packages names 1`] = `
[
  "react-native-error-boundary",
  "gitmoji-cli",
  "gitmojis",
  "react-native-layout-debug",
  "hyper-materialshell",
  "generator-starterkit",
]
`;
